Arp

kgrubb opened the pull request:
__CHANGES:__

* implement all of arp get/set
* add tests for each field

kgrubb committed:
improve arp implementation with all header fields

kgrubb committed:
swap from u16 to 2-byte array for hardware_type, protocol_type, and operation

kgrubb committed:
use original var names

nathaniel-d-ef commented on src/arp.rs:
```diff
@@ -18,4 +18,216 @@ pub struct ArpHdr {
+    // Returns the sender hardware address field.
+    #[inline]
+    pub fn sha(&self) -> [u8; 6] {
+        self.sha
```
Aren't sha, spa, tha, and tpa variable length?

svencowart replied:
From Gemini:

> For the vast majority of practical applications, the Hardware Length in an ARP header is 6, and the Protocol Length is 4.
> 
> This is because the most common use of the Address Resolution Protocol (ARP) is to resolve IPv4 addresses (the protocol) to Ethernet MAC addresses (the hardware).
> 
> ### A Closer Look at the Fields
> Hardware Length (HLEN): This 1-byte field specifies the length of the hardware address. For Ethernet, the MAC address is 48 bits, which is equal to 6 bytes.
> Protocol Length (PLEN): This 1-byte field specifies the length of the logical address. for IPv4, the address is 32 bits, which is equal to 4 bytes.
> The design of the ARP protocol is flexible and can theoretically be used with other network protocols and hardware types that have different address lengths. However, its primary and overwhelmingly common implementation is for IPv4 over Ethernet. Therefore, in practice, you will almost always find these specific values in an ARP header.

This verifies my research. I couldn't find any practical examples of using ARP for anything other than Ethernet and IPv4. Our own PCAPs demonstrate that. I'd say, let's assume that is the case, and therefore we don't need to make these fields variable length. It would be good to make a related call out on those fields in the struct within a doc comment.

svencowart commented on src/arp.rs:
```diff
@@ -18,4 +18,216 @@ pub struct ArpHdr {
+            tha: [0; 6],
+            tpa: [0; 4],
+        }
+    }
```
You could add a default function to the struct itself

svencowart commented on src/arp.rs:
```diff
@@ -18,4 +18,216 @@ pub struct ArpHdr {
     pub const LEN: usize = mem::size_of::<ArpHdr>();
+
+    // Returns the hardware type field.
+    #[inline]
```
For all these functions and the ArpHdr, can you ask AI to write doc comments

kgrubb committed:
add rustdoc information. add a default function

jecsand838 committed:
Update vxlan.rs (#58)

Introduce robust and RFC-compliant structures for Virtual eXtensible Local Area Network (RFC 7348) and headers, significantly improving VXLAN message handling.

Key improvements include:

* Adding vni_and_reserved2 Attribute
* Removed BitfieldUnit
* Added implementation method
* Added Unit Tests

Co-authored-by: Sven Cowart <svencowart@gmail.com>

veronica-m-ef committed:
Adds support for GENEVE (#67)

*   **`GeneveHdr` struct**: A `#[repr(C, packed)]` struct is defined to accurately model the Geneve header format, ensuring correct memory layout.
*   **Field Accessors**: Provides `get` and `set` methods for all fields of the Geneve header, including:
    *   `ver` (Version)
    *   `opt_len` (Option Length)
    *   `o_flag` (OAM Flag)
    *   `c_flag` (Critical Options Present Flag)
    *   `protocol_type` (Encapsulated Protocol Type, handled with network byte order)
    *   `vni` (Virtual Network Identifier, handled with network byte order and 24-bit masking)
*   **Bitfield Handling**: The methods correctly handle bitfield extraction and insertion for `ver`, `opt_len`, `o_flag`, and `c_flag`.
*   **Network Byte Order**: `protocol_type` and `vni` conversions correctly handle big-endian (network byte order).
*   **Unit Tests**: Comprehensive unit tests are included to verify the correct behavior of all getters, setters, bitfield operations, and byte order conversions.

This addition provides foundational support for parsing and constructing Geneve encapsulated packets within the project.

vadorovsky committed:
chore: Fix rustfmt errors (#70)

kgrubb committed:
test: Multi-target CI Stages (#62)

* Update `actions/checkout` action.
* Use `dtolnay/rust-toolchain` action instead of `actions-rs`.
* Run rustfmt and clippy.
* Add cross targets.

Co-authored-by: Michal Rostecki <vadorovsky@disroot.org>

kgrubb committed:
Apply suggestions from code review

Co-authored-by: Michal Rostecki <vadorovsky@disroot.org>

kgrubb committed:
derive default instead of explicitly creating it. import memoffset and use integer types in getters and setters if the given field is an integer, represented as an array.

kgrubb committed:
fix formatting in arp
